src/goto-instrument/value_set_fi_fp_removal.cpp

```diff
@@ -0,0 +1,250 @@
+  {
+    goto_programt::targett t = new_code_gotos.add_instruction();
+    t->make_assertion(false_exprt());
+    t->source_location.set_property_class("pointer dereference");
+    t->source_location.set_comment("invalid function pointer");
+  }
+
+  goto_programt new_code;
+
+  // patch them all together
+  new_code.destructive_append(new_code_gotos);
+  new_code.destructive_append(new_code_calls);
+  new_code.destructive_append(final_skip);
+
+  // set locations
+  Forall_goto_program_instructions(it, new_code)
+  {
+    irep_idt property_class = it->source_location.get_property_class();
+    irep_idt comment = it->source_location.get_comment();
+    it->source_location = target->source_location;
+    if(!property_class.empty())
+      it->source_location.set_property_class(property_class);
+    if(!comment.empty())
+      it->source_location.set_comment(comment);
+  }
+
+  goto_programt::targett next_target = target;
+  next_target++;
+
+  goto_program.destructive_insert(next_target, new_code);
+
+  // We preserve the original dereferencing to possibly catch
+  // further pointer-related errors.
+  code_expressiont code_expression(function);
+  code_expression.add_source_location() = function.source_location();
+  target->code.swap(code_expression);
+  target->type = OTHER;
+}
+
+void value_set_fi_fp_removal(goto_modelt &goto_model)
+{
+  std::cout << "Doing FI value set analysis\n";
+
+  const namespacet ns(goto_model.symbol_table);
+  value_set_analysis_fit value_sets(ns);
+  value_sets(goto_model.goto_functions);
+
+  std::cout << "Instrumenting\n";
+
+  // now replace aliases by addresses
+  for(auto &f : goto_model.goto_functions.function_map)
+  {
+    for(auto target = f.second.body.instructions.begin();
+        target != f.second.body.instructions.end();
+        target++)
+    {
+      if(target->is_function_call())
+      {
+        const auto &call = to_code_function_call(target->code);
+        if(call.function().id() == ID_dereference)
+        {
+          std::cout << "CALL at " << target->source_location << ":\n";
+
+          const auto &pointer = to_dereference_expr(call.function()).pointer();
+          std::list<exprt> addresses;
+          value_sets.get_values(f.first, target, pointer, addresses);
+
+          std::set<symbol_exprt> functions;
+
+          for(const auto &address : addresses)
+          {
+            // is this a plain function address?
+            // strip leading '&'
+            if(address.id() == ID_object_descriptor)
+            {
+              const auto &od = to_object_descriptor_expr(address);
+              const auto &object = od.object();
+              if(object.type().id() == ID_code && object.id() == ID_symbol)
+                functions.insert(to_symbol_expr(object));
+            }
+          }
+
+          for(const auto &f : functions)
+            std::cout << "  function: " << f.get_identifier() << '\n';
+
+          if(functions.size() > 0)
+            remove_function_pointer(
```

Comment:
So if it has no ideas, we leave a function pointer behind? That seems problematic for downstream passes, probably better to turn it into an assume

Reply:
What do you mean turn it into an assume? 

The standard remove function pointers pass is run after this pass, so leaving the function pointer behind seems like a sensible thing to do because we then default to the standard behaviour.

Reply:
Probably an assume false, but the default is fine too; in that case in the doc comment for the pass mention that it can leave function pointers behind and some other removal pass must run after it.

Reply:
assume false would potentially hide bugs

Reply:
Sorry assert, but the fall-through is fine, happy to stick with that so long as the pass is documented as needing a subsequent cleanup pass to take care of residual function pointers